relayer/chains/solana/listener.go

```diff
@@ -28,19 +29,222 @@ func (p *Provider) Listener(ctx context.Context, lastSavedHeight uint64, blockIn
 	// 	p.log.Error("failed to init xcall", zap.Error(err))
 	// }
 
+	// return nil
+
 	if err := p.SendMessage(ctx, &relayertypes.Message{
 		Dst:  "0x3.icon",
 		Data: []byte("hello"),
 	}); err != nil {
 		p.log.Error("failed to send message", zap.Error(err))
 	}
 
-	p.log.Info("started querying from height", zap.Uint64("height", startHeight))
+	// fromSignature := "2D1RxZptfGrfHwqfzMD3Z6TdnGb5Ugff3ZHuDmNT7xMTfpRtaBSpdwrC2R8qrioDDFvkU3TU5yTSukv2iByoGcuN"
+	fromSignature := ""
+
+	p.log.Info("started querying from height", zap.String("from-signature", fromSignature))
+
+	return p.listenByPolling(ctx, fromSignature, blockInfo)
+}
+
+func (p *Provider) listenByPolling(ctx context.Context, fromSignature string, blockInfo chan *relayertypes.BlockInfo) error {
+	ticker := time.NewTicker(3 * time.Second)
+
+	if fromSignature != "" {
+		fromSign, err := solana.SignatureFromBase58(fromSignature)
+		if err != nil {
+			return err
+		}
+		if err := p.processTxSignature(ctx, fromSign, blockInfo); err != nil {
+			p.log.Error("failed to process tx signature", zap.String("signature", fromSign.String()), zap.Error(err))
+		}
+	}
 
 	for {
 		select {
 		case <-ctx.Done():
 			return ctx.Err()
+		case <-ticker.C:
+			//fetch txSigns from most recent to oldest. 0th index is the most recent and last index is oldest
+			txSigns, err := p.getSignatures(ctx, fromSignature)
+			if err != nil {
+				p.log.Error("failed to get signatures", zap.Error(err))
+				break
+			}
+			if len(txSigns) > 0 {
+				//next query start from most recent signature.
+				fromSignature = txSigns[0].Signature.String()
+			}
+			//start processing from last index i.e oldest signature
+			for i := len(txSigns) - 1; i >= 0; i-- {
+				sign := txSigns[i].Signature
+				if err := p.processTxSignature(ctx, sign, blockInfo); err != nil {
+					p.log.Error("failed to process tx signature", zap.String("signature", sign.String()), zap.Error(err))
+				}
+			}
+		}
+	}
+}
+
+func (p *Provider) processTxSignature(ctx context.Context, sign solana.Signature, blockInfo chan *relayertypes.BlockInfo) error {
+	txn, err := p.client.GetTransaction(ctx, sign, nil)
+	if err != nil {
+		return fmt.Errorf("failed to get txn with sign %s: %w", sign, err)
+	}
+
+	if txn.Meta != nil && len(txn.Meta.LogMessages) > 0 {
+		event := types.SolEvent{Slot: txn.Slot, Signature: sign, Logs: txn.Meta.LogMessages}
+		messages, err := p.parseMessagesFromEvent(event)
+		if err != nil {
+			return fmt.Errorf("failed to parse messages from event [%+v]: %w", event, err)
+		}
+		if len(messages) > 0 {
```

Comment:
Isn't this check redundant ?

Reply:
> Isn't this check redundant ?

Uhh, this check is just to make sure, we don't send blockInfo with empty messages in the buffer. This has several advantages like we are not populating buffer with useless info. And since it is likely for events logs to have messages that we don't need, there is good probability to occur this situation of empty messages, and if we keep pushing info with empty message in the buffer, we are just wasting slot of buffer.